File: python/backtracking/nqueues_ii.py

# N皇后问题II

class Solution:

    shu, pie, na = set(), set(), set()
    res = 0

    # 传统的解法
    def totalNQueens1(self, n: int) -> int:
        self.dfs(n, 0)
        return self.res

    def dfs(self, n, i):
        if i == n:
            self.res += 1
            return
        
        for j in range(n):
            if j not in self.shu and i - j not in self.pie and i + j not in self.na:
                self.shu.add(j), self.pie.add(i - j), self.na.add(i + j)
                self.dfs(n, i + 1)
                self.shu.remove(j), self.pie.remove(i - j), self.na.remove(i + j)


    # 使用位运算
    def totalNQueens2(self, n: int) -> int:
        self.count = 0

    def dfs2(self, n, row, shu, pie, na):
        if row == n:
            self.count += 1
            return
        

File: go/backtracking/nqueues_ii.go
package main

//N皇后问题II

//常规解法

func totalNQueens(n int) int {
	shu, pie, na := make(map[int]bool), make(map[int]bool), make(map[int]bool)
	var dfs func(n, i int, shu, pie, na map[int]bool)
	res := 0
	dfs = func(n, i int, shu, pie, na map[int]bool) {
		if i == n {
			res++
			return
		}
		for j := 0; j < n; j++ {
			if !shu[j] && !pie[i-j] && !na[i+j] {
				shu[j], pie[i-j], na[i+j] = true, true, true
				dfs(n, i+1, shu, pie, na)
				shu[j], pie[i-j], na[i+j] = false, false, false
			}
		}
	}
	dfs(n, 0, shu, pie, na)
	return res
}

//位运算
func totalNQueens2(n int) (ans int) {
	var solve func(row, columns, diagonals1, diagonals2 int)
	solve = func(row, columns, diagonals1, diagonals2 int) {
		if row == n {
			ans++
			return
		}
		availablePositions := (1<<n - 1) &^ (columns | diagonals1 | diagonals2)
		for availablePositions > 0 {
			position := availablePositions & -availablePositions
			solve(row+1, columns|position, (diagonals1|position)<<1, (diagonals2|position)>>1)
			availablePositions &^= position
		}
	}
	solve(0, 0, 0, 0)
	return
}
